Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+            if (ancestorClass == BooleanObject::info()) {
+                recordFailure("wrapped boolean"_s);
+                return;
+            }
+            if (ancestorClass == BigIntObject::info()) {
+                recordFailure("wrapped big int"_s);
+                return;
+            }
+        } while ((ancestorClass = ancestorClass->parentClass));
+        auto& structure = *object.structure();
+        if (object.structureID() != structure.id()) {
+            recordFailure("structure ID mismatch"_s);
+            return;
+        }
+        if (structure.hasPolyProto()) {
+            recordFailure("hasPolyProto"_s);
+            return;
+        }
+        if (structure.storedPrototype() != m_globalObject.objectPrototype()) {
+            recordFailure("non-standard object prototype"_s);
+            return;
+        }
+        if (!m_checkedObjectPrototype) {
+            if (mayHaveToJSON(*m_globalObject.objectPrototype())) {
+                recordFailure("object prototype may have toJSON"_s);
+                return;
+            }
+            m_checkedObjectPrototype = true;
+        }
+        if (!remainingCapacity()) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length++] = '{';
+        if (canPerformFastPropertyEnumerationForJSONStringify(&structure)) {
+            structure.forEachProperty(m_vm, [&](const PropertyTableEntry& entry) -> bool {
+                if (entry.attributes() & PropertyAttribute::DontEnum)
+                    return true;
+                auto& name = *entry.key();
+                if (name.isSymbol()) {
+                    recordFailure("symbol"_s);
+                    return false;
+                }
+                if (!name.is8Bit()) {
+                    recordFailure("16-bit property name"_s);
+                    return false;
+                }
+                bool needComma = m_buffer[m_length - 1] != '{';
+                unsigned nameLength = name.length();
+                if (remainingCapacity() < needComma + 1 + nameLength + 2) {
+                    recordBufferFull();
+                    return false;
+                }
+                if (needComma)
+                    m_buffer[m_length++] = ',';
+                m_buffer[m_length] = '"';
+                auto* characters = name.characters8();
+                for (unsigned i = 0; i < nameLength; ++i) {
+                    auto character = characters[i];
+                    if (WTF::escapedFormsForJSON[character]) {
+                        recordFailure("property name character needs escaping"_s);
+                        return false;
+                    }
+                    m_buffer[m_length + 1 + i] = character;
+                }
+                m_buffer[m_length + 1 + nameLength] = '"';
+                m_buffer[m_length + 1 + nameLength + 1] = ':';
+                m_length += 1 + nameLength + 2;
+                append(object.getDirect(entry.offset()));
```

Comment:
Even though we do not see non-user-observable behavior as no-side-effect, still it is possible that structure can be transitioned. For example, `append` will call `getOwnPropertyNames`, and it could change the state of this object.
To avoid crash (like, object's structure could be flattened, and in that case, it is possible that offset->value mapping is completely changed in the middle of iteration), we should buffer values in some places, and use them.
See `Object.assign`'s implementation for more detail.

Reply:
I’m assuming the answer is no. Please let me know if I am wrong!

Reply:
We’re discussing this below, so I will resolve this one.